### ISSUE TITLE
proposal: Simplify credentialset matching

### ISSUE THREAD

**Issue by technosophos** [cnabio/duffle#266]
*tl;dr* There might be an easier way to re-use credentials than what we have now. To do this, we'd treat credential _names_ as important.

## Right now: Matching is done by comparing destinations

Right now, a credentialset looks like this:

```yaml
name: az
credentials:
  - name: subscription_id
    source:
      value: SOMETHING
    destination:
      env: AZURE_SUBSCRIPTION_ID
  - name: tenant_id
    source:
      value: SOMETHING
    destination:
      env: AZURE_TENANT_ID
  - name: client_id
    source:
      value: SOMETHING
    destination:
      env: AZURE_CLIENT_ID
  - name: client_secret
    source:
      value: SOMETHING
    destination:
      env: AZURE_CLIENT_SECRET
```

This declares three things:

- a *source*
- a *destination*
- a *name*

And _name_ is basically unused

Now a `bundle.json` declares its credentials like this:

```json
"credentials": {
        "tenant_id": {
            "env": "AZURE_TENANT_ID"
        },
        "client_id": {
            "env": "AZURE_CLIENT_ID"
        },
        "client_secret": {
            "env": "AZURE_CLIENT_SECRET"
        },
        "subscription_id": {
            "env": "AZURE_SUBSCRIPTION_ID"
        }
    }
```

Name is unused.

So in the examples above, the `credentialset` satisfies the `bundle.json`'s credential requirement _BECAUSE_ The names of the environment variables match up.

## But wait...

Here's the problem: We can look at another bundle and see that it requires _the same credentials_, but because of the names of the env vars, we'd have to write a different credentialset. And the names of the env vars are dictated by the internals of Terraform (in this case. Ansible is similar).

```json
"credentials": {
        "tenant_id": {
            "env": "TF_VAR_tenant_id"
        },
        "client_id": {
            "env": "TF_VAR_client_id"
        },
        "client_secret": {
            "env": "TF_VAR_client_secret"
        },
        "subscription_id": {
            "env": "TF_VAR_subscription_id"
        },
        "ssh_authorized_key": {
            "env": "TF_VAR_ssh_authorized_key"
        }
    }
```

Now the credential set we provided *does not satisfy these requirements* even though the credentials actually _would_ satisfy the requirements. The reason is that our credentialset env vars are named differently than our terraform env vars.

Ideally, the same credentialset should be usable for both bundles, since it contains the credentials for both bundles.

## The Solution: Match on `name` instead of on the env var or the path

Credentialsets are named collections of source->dest mappings.

The `credentials` field in `bundle.json` is a named set of destinations.

_Why not map credentials by name instead of path?_

In this case, the credentialset would merely be _named list of local credentials_ and the `credentials` section of `bundle.json` would be a _named list of destinations_:

Credentialset:

```yaml
name: az
credentials:
  - name: subscription_id
    source:
      value: SOMETHING
  - name: tenant_id
    source:
      value: SOMETHING
  - name: client_id
    source:
      value: SOMETHING
  - name: client_secret
    source:
      value: SOMETHING
```

Credential lists remain the same:
```json
"credentials": {
        "tenant_id": {
            "env": "TF_VAR_tenant_id"
        },
        "client_id": {
            "env": "TF_VAR_client_id"
        },
        "client_secret": {
            "env": "TF_VAR_client_secret"
        },
        "subscription_id": {
            "env": "TF_VAR_subscription_id"
        },
        "ssh_authorized_key": {
            "env": "TF_VAR_ssh_authorized_key"
        }
    }
```

And the matching algorithm compares the _name_ field in the credentialset to the _name_ in the `credentials` declaration.

This accomplishes three things:

1. It simplifies credentialsets
2. It removes the need for the Duffle user to know where the credentials have to be mounted inside of the invocationImage
3. it makes credentials much easier to map/reuse across multiple bundles
